stl/inc/xutility

```diff
@@ -3063,8 +3063,477 @@ namespace ranges {
+        && convertible_to<_Second, tuple_element_t<1, _Ty>>;
+
+    template <input_or_output_iterator _It, sentinel_for<_It> _Se = _It,
+        subrange_kind _Ki = sized_sentinel_for<_Se, _It> ? subrange_kind::sized : subrange_kind::unsized>
+        requires (_Ki == subrange_kind::sized || !sized_sentinel_for<_Se, _It>)
+    class subrange;
+    // clang-format on
+
+    template <class _It, class _Se, subrange_kind _Ki,
+        bool _Store = _Ki == subrange_kind::sized && !sized_sentinel_for<_Se, _It>>
+    class _Subrange_base : public view_interface<subrange<_It, _Se, _Ki>> { // TRANSITION, [[no_unique_address]]
+    protected:
+        using _Size_type = _Make_unsigned_like_t<iter_difference_t<_It>>;
+
+        static constexpr bool _Store_size = true;
+        _Size_type _Size                  = 0;
+
+    public:
+        _Subrange_base() = default;
+        constexpr _Subrange_base(const _Size_type& _Size_) noexcept : _Size(_Size_) {}
+    };
+
+    template <class _It, class _Se, subrange_kind _Ki>
+    class _Subrange_base<_It, _Se, _Ki, false> : public view_interface<subrange<_It, _Se, _Ki>> {
+    protected:
+        using _Size_type = _Make_unsigned_like_t<iter_difference_t<_It>>;
+
+        static constexpr bool _Store_size = false;
+
+    public:
+        _Subrange_base() = default;
+        constexpr _Subrange_base(const _Size_type&) noexcept {}
+    };
+
+    // clang-format off
+    template <input_or_output_iterator _It, sentinel_for<_It> _Se, subrange_kind _Ki>
+        requires (_Ki == subrange_kind::sized || !sized_sentinel_for<_Se, _It>)
+    class subrange : public _Subrange_base<_It, _Se, _Ki> {
+    private:
+        using _Mybase = _Subrange_base<_It, _Se, _Ki>;
+        using typename _Mybase::_Size_type;
+        using _Mybase::_Store_size;
+
+        // TRANSITION, [[no_unique_address]]:
+        /* [[no_unique_address]] */ _It _First{};
+        /* [[no_unique_address]] */ _Se _Last{};
+        // [[no_unique_address]] conditional_t<_Store_size, _Size_type, _Nil> _Size{};
+
+        template <class _Rng>
+        constexpr subrange(true_type, _Rng&& _Val) : subrange{_STD forward<_Rng>(_Val), _RANGES size(_Val)} {
+            // delegation target for subrange(_Rng&&) when _Rng models sized_range
+        }
+        template <class _Rng>
+        constexpr subrange(false_type, _Rng&& _Val) : subrange{_RANGES begin(_Val), _RANGES end(_Val)} {
+            // delegation target for subrange(_Rng&&) when _Rng does not model sized_range
+        }
+
+    public:
+        subrange() = default;
+
+        template <_Convertible_to_non_slicing<_It> _It2>
+        constexpr subrange(_It2 _First_, _Se _Last_) requires (!_Store_size)
+            : _First(_STD move(_First_)), _Last(_STD move(_Last_)) {}
+
+        template <_Convertible_to_non_slicing<_It> _It2>
+        constexpr subrange(_It2 _First_, _Se _Last_, const _Size_type _Size_) requires (_Ki == subrange_kind::sized)

```

Comment:
Possibly there should, but I couldn't bring myself to force a copy for the constructor argument to the specialization of `_Subrange_base` that simply drops the argument on the floor given that there will eventually be ADT size types. The `_Subrange_base` constructors are consistent with each other rather than with this constructor.

I note that this wouldn't be an issue at all if we had `[[no_unique_address]]`.

Reply:
Ok.